bmorph/util/mizuroute_utils.py

```diff
@@ -337,11 +337,90 @@ def map_var_to_segs(routed: xr.Dataset, map_var: xr.DataArray, var_label: str,
+                            site_seg_mapping: dict, met_hru: xr.Dataset=xr.Dataset(),
+                            route_var: str='IRFroutedRunoff'):
+    '''
+    Prepare mizuroute output for bias correction via the blendmorph algorithm. This
+    allows an optional dataset of hru meteorological data to be given for conditional
+    bias correction.
+
+    Parameters
+    ----------
+    topo:
+        Topology dataset for running mizuRoute.
+        We expect this to have ``seg`` and ``hru`` dimensions
+    routed:
+        The initially routed dataset from mizuRoute
+    reference:
+        A dataset containing reference flows for bias correction.
+        We expect this to have ``site`` and ``time`` dimensions.
+    site_seg_mapping:
+        A lookup table from the ``reference.site`` to ``topo.seg``
+    met_hru:
+        (Optional) A dataset of meteorological data to be mapped
+        onto the stream segments to facilitate conditioning.
+        All variables in this dataset will automatically be mapped
+        onto the stream segments and returned
+    route_var:
+        Name of the variable of the routed runoff in the ``routed``
+        dataset. Defaults to ``IRFroutedRunoff``
+
+    Returns
+    -------
+    met_seg:
+        A dataset with the required data for applying the ``blendmorph``
+        routines. See the ``blendmorph`` documentation for further information.
+    '''
+    # Provide some convenience data for mapping/loops
+    ref_sites = list(site_seg_mapping.keys())
+    ref_segs = list(site_seg_mapping.values())
+    hru_2_seg = topo['seg_hru_id'].values
+    met_vars = set(met_hru.variables.keys()) - set(met_hru.coords)
+
+    # Prep met data structures
+    met_seg = xr.Dataset()
+    for v  in met_vars:
+        met_seg[v] = xr.DataArray(data=np.nan, dims=('time', 'seg', ),
+                            coords={'time': met_hru['time'], 'seg': topo['seg']})
+
+    # Map from hru -> segment for met data
+    # In case a mapping doesn't exist to all segments,
+    # we define a neighborhood search to spatially average
+    null_neighborhood = [-3, -2, -1, 0, 1, 2, 3]
+    for var in met_vars:
+        for seg in met_seg['seg'].values:
+            subset = np.argwhere(hru_2_seg == seg).flatten()
+            # First fallback, search in the null_neighborhood
+            if not len(subset):
+                subset = np.hstack([np.argwhere(hru_2_seg == seg-offset).flatten()
+                                    for offset in null_neighborhood])
+            # Second fallback, use domain average
+            if not len(subset):
+                subset = met_hru['hru'].values
+            met_seg[var].loc[{'seg': seg}] = met_hru[var].isel(hru=subset).mean(dim='hru')
+
+    # Get the longest overlapping time period between all datasets
+    [routed, reference, met_seg] = trim_time([routed, reference, met_seg])
+    routed = calculate_blend_vars(routed, topo, reference)
```

Comment:
To allow sub-setting of reference sites, `ref_sites` from `site_seg_mapping` should be passed in here, otherwise all the sites in reference will be automatically used in preparation for blending.

Reply:
This very well could be where my issues crop up, but there is no otherway to axis that parameter currently

Reply:
I was approaching this different than you were it seems. If you check my original code snippet when I open the `reference` I'm already selecting out the `ref_sites`, which subsets this all out early. After looking at this and thinking about it, it seems that this approach makes sense, and also simplifies things. I've updated the code so that we don't need to pass in the `site_seg_mapping` at all. A user should just select the sites they want to train on out of the reference dataset up front and then everything else follows.